# <span style="color:green"> Start Here </span> 

[Please Complete the following form with your details]

Student Name - Alexandra Pinto<br>
Student id - 20211599<br>
Contact e-mail - 20211599@novaims.unl.pt<br>


Student Name - Marco Galão<br>
Student id - 20201545<br>
Contact e-mail - r20201545@novaims.unl.pt<br>

# <span style="color:brown"> Part 1 -  Variable Declaration and Manipulation </span> 

## Exercise I - of Lists and Random numbers

**1**: Declare a variable X that stores a list of 100 integers randomly sampled between -100 and 100.<br>
Note: You are not allowed to use third party libraries -- such as Numpy, Pandas, or Scipy -- in this exercise.

In [1]:
# Import the random module
import random

In [2]:
# Create a list X with 100 random integers between -100 and 100 (inclusive)
X = [random.randint(-100, 100) for i in range(100)]

# Explanation:
# random.randint(-100, 100) generates a random integer between -100 and 100 (inclusive)
# range(100) ensures that 100 random integers are generated and added to the list X

In [3]:
# Print the list X
print("X:", X)

X: [-38, -38, -72, 7, 65, 81, 33, 48, 75, -88, 3, -39, -12, -69, 93, -41, 64, -63, -66, 39, 66, 94, 35, -86, -3, 33, -37, 29, -15, -24, 23, 22, -41, -41, -15, -15, -60, 63, 19, -9, -44, 25, 11, 75, -82, 5, -31, 82, -81, -18, -60, 5, 30, -26, -55, 51, -99, -73, 17, -28, 38, 37, -97, 31, -52, -27, 15, -92, -67, -81, -23, -87, 89, 81, -5, -87, 25, 27, -42, -21, 83, -25, 57, 74, -78, 13, -21, 57, 15, -20, 94, -64, -95, -13, 79, 21, 39, 89, 36, 75]


**2**: How many odd numbers are in list X?


In [4]:
# Initialize a counter to keep track of the odd numbers
odd_count = 0

# Iterate through each number in the list X to check if it's odd
for i in X:
    # If the number is odd (there is a remainder when divided by 2), increment the odd number counter by 1
    if i % 2 != 0:
        odd_count += 1

In [5]:
# Print the answer
print(f"There are {odd_count} odd numbers in list X.")

There are 68 odd numbers in list X.


**2.1** Check if there are as many even numbers as there are odd numbers, else discard the list and generate a new one.

In [6]:
# In this exercise, we will adapt the code from exercise 2

# Initialize a counter to keep track of the even numbers
even_count = 0

# Iterate through each number in the list X to check if it's even
for i in X:
    # If the number is even (there is no remainder when divided by 2), increment the even number counter by 1
    if i % 2 == 0:
        even_count += 1

In [7]:
# Print the answer
print(f"There are {even_count} odd numbers in list X.")

There are 32 odd numbers in list X.


In [8]:
# Now we will compare the counts of odd and even numbers

# Check if the number of odd and even numbers are equal or different
if odd_count == even_count:
    print("The list X has an equal number of odd and even numbers.")
else:
    print("The list X has a different number of odd and even numbers, so we will generate a new list X.")
    
    # Use the code from exercise 1
    # Regenerate the list X with 100 random integers between -100 and 100 (inclusive)
    X = [random.randint(-100, 100) for i in range(100)]
    
    # Explanation:
    # random.randint(-100, 100) generates a random integer between -100 and 100, including both
    # range(100) ensures that 100 random integers are generated and added to the list X

    # Print X
    print("X:", X)

The list X has a different number of odd and even numbers, so we will generate a new list X.
X: [13, 33, -37, -33, 52, -79, -70, -67, 85, 50, 90, 81, -96, -55, -11, 7, 3, 56, 50, -15, 43, 6, -65, -72, -76, 95, 73, 40, -46, 40, 87, -14, 43, 13, 44, 36, -11, -43, 62, -96, 87, 67, 7, 91, -37, 30, -74, 36, 9, 35, -20, -35, -23, 86, 96, 31, 20, 81, 41, 97, -33, 61, 10, 25, 48, 57, 40, -55, -38, 98, -64, -59, -54, -47, -89, 99, 85, 52, 18, 51, 5, -90, -44, -64, -12, -13, 52, -98, 85, -68, 8, -43, 77, -23, 7, -35, 48, -23, 60, 31]


**2.2** Create a pipeline to automatically perform the tasks above such that you avoid having to run the cells multiple times to reach the desired outcome.

In [9]:
# Import the random module (to generate random integers)
import random

In [10]:
# Define the function GenerateList
def GenerateList():
    """
    Generates a list of 100 random integers between -100 and 100 (inclusive).

    Returns:
        list: A list of 100 random integers between -100 and 100 (inclusive).
    """
    
    return [random.randint(-100, 100) for i in range(100)]
    # Explanation:
    # random.randint(-100, 100) generates a random integer between -100 and 100 (inclusive)
    # range(100) ensures that 100 random integers are generated and added to the list X

# Define the function CountEvenOddNumbers
def CountEvenOddNumbers(X: list) -> tuple:
    """
    Counts the number of even and odd numbers in a list of numeric types.

    Args:
    - X (list): A list of numeric types.

    Returns:
    - tuple: A tuple containing two integers - (even_count, odd_count) where:
             - even_count is the number of even integers in the list.
             - odd_count is the number of odd integers in the list.
    """

    # Initialize counters to keep track of even and odd numbers
    even_count = 0
    odd_count = 0

    # Iterate through each number in the list X to check if it's even or odd
    for i in X:
        # If the number is even (there is no remainder when divided by 2), increment the even number counter by 1
        if i % 2 == 0:
            even_count += 1  
        # Otherwise, it's odd, so increment the odd number counter by 1
        else:
            odd_count += 1
    
    # Return the counts of even and odd numbers as a tuple
    return even_count, odd_count

# Define the function Pipeline
def Pipeline():
    """
    Generates a list of 100 random integers between -100 and 100 (inclusive) and
    checks if it has an equal number of even and odd numbers.
    If not, it regenerates the list until the counts of even and odd numbers are equal.

    Returns:
    - None
    """

    # Generate the initial list X
    X = GenerateList()
    
    # Calculate the initial counts of even and odd numbers
    even_count, odd_count = CountEvenOddNumbers(X)
    
    # Continue regenerating list X until the number of even and odd numbers are equal
    while even_count != odd_count:
        print("The list X has a different number of odd and even numbers, so we will generate a new list X.")
        
        # Regenerate a new list X
        X = GenerateList()

        # Print the regenerated list X
        print("X:", X, "\n")

        # Recalculate the counts of even and odd numbers
        even_count, odd_count = CountEvenOddNumbers(X)
        
    # When the loop breaks, it means the number of even and odd numbers is equal
    print("The list X has equal number of even and odd numbers.")
    
    # Print the final list X
    print("Final X:", X)

In [11]:
# Call the function Pipeline
Pipeline()

The list X has a different number of odd and even numbers, so we will generate a new list X.
X: [-81, -57, 52, -58, -52, 0, -54, -10, -62, -39, 66, 46, -90, 15, 14, 41, 4, 96, -91, 63, 30, -79, -56, 11, 41, 2, -83, 25, 4, 17, -80, -92, 100, -58, 59, 39, 32, -17, 56, -20, 46, -79, 16, 97, -5, 32, -25, -90, -13, 70, 78, -12, -73, -95, 32, 93, -9, -57, -88, 100, -70, -13, 15, 24, -86, -80, -40, 93, -64, 85, -49, 90, -63, 99, -68, 18, -2, -57, -94, 13, -51, 37, 27, 25, 53, -35, -61, 11, 62, 12, 50, 53, 12, 33, 58, -9, -16, 26, -88, 3] 

The list X has a different number of odd and even numbers, so we will generate a new list X.
X: [-81, -72, -32, -26, -1, 53, 70, -22, 9, -43, 3, -55, -63, 2, -9, 94, -52, 35, -14, -61, 61, 17, -64, -70, -22, -37, 73, -9, -43, 91, 39, -70, -65, -15, -57, -39, 16, -15, 82, 10, 95, -69, -97, 87, -22, 32, -26, 72, 31, 46, 24, 16, 94, 84, -13, 89, 83, 92, -5, 76, 44, 92, -95, 88, 82, 62, 21, 63, -15, 75, 29, -6, 9, 57, 7, -80, -46, 13, 96, -11, -52, -26, -38, -6

**3**: Print the number of digits that the 5th and 100th element of the list have. <br>
*Note: For instance, the number 1 contains one digit, the number 10 contains two digits, the number -2 contains one digit.*

In [12]:
# Access the fifth and hundredth elements of the list X (indexing starts at 0)
fifth_element = X[4]
hundredth_element = X[99]

In [13]:
# To determine the number of digits in a number, we will use a combination of three functions:
# - abs: Returns the absolute value of the number to ensure we count digits correctly, regardless of sign
# - str: Converts the absolute value to a string, allowing us to count its length
# - len: Returns the length of the string representation, which corresponds to the number of digits

# Print the answer
print(f"The 5th element is {fifth_element} and has {len(str(abs(fifth_element)))} digits.")
print(f"The 100th element is {hundredth_element} and has {len(str(abs(hundredth_element)))} digits.")

The 5th element is 52 and has 2 digits.
The 100th element is 31 and has 2 digits.


**4**: Is the sum total of all the numbers in the list even or odd?

In [14]:
# Store the sum of all numbers in the list
total_sum = sum(X)

# Determine if the total sum is even or odd and print the answer
if total_sum % 2 == 0:
    print(f"The sum of all the numbers in list X is even and is equal to {total_sum}.")
else:
    print(f"The sum of all the numbers in list X is odd and is equal to {total_sum}.")

The sum of all the numbers in list X is even and is equal to 906.


**5**: What is the average of the list X?

In [15]:
# Calculate the average of the list X
avg = sum(X) / len(X)

In [16]:
# Print the answer
print(f"The average of the list X is {avg}.")

The average of the list X is 9.06.


**5.1.** What is the population standard deviation? <br>

In [17]:
# Calculate the squared differences from the average
squared_diffs = [(x - avg) ** 2 for x in X]

# Calculate the population variance
pop_var = sum(squared_diffs) / len(X)

# Calculate the population standard deviation
pop_std_dev = pop_var ** 0.5

In [18]:
# Print the answer
print(f"The population standard deviation of X list is {round(pop_std_dev, 2)}.")

The population standard deviation of X list is 56.17.


**6**: Sort list X in descending order and store the result in variable Xsort. <br>

In [19]:
# Sort the list X in descending order using the sorted() function
# The reverse=True argument ensures that the sorting is done in descending order
Xsort = sorted(X, reverse=True)

In [20]:
# Print the original list X and the sorted list Xsort for comparison
print("X:", X)
print("Xsort:", Xsort)

X: [13, 33, -37, -33, 52, -79, -70, -67, 85, 50, 90, 81, -96, -55, -11, 7, 3, 56, 50, -15, 43, 6, -65, -72, -76, 95, 73, 40, -46, 40, 87, -14, 43, 13, 44, 36, -11, -43, 62, -96, 87, 67, 7, 91, -37, 30, -74, 36, 9, 35, -20, -35, -23, 86, 96, 31, 20, 81, 41, 97, -33, 61, 10, 25, 48, 57, 40, -55, -38, 98, -64, -59, -54, -47, -89, 99, 85, 52, 18, 51, 5, -90, -44, -64, -12, -13, 52, -98, 85, -68, 8, -43, 77, -23, 7, -35, 48, -23, 60, 31]
Xsort: [99, 98, 97, 96, 95, 91, 90, 87, 87, 86, 85, 85, 85, 81, 81, 77, 73, 67, 62, 61, 60, 57, 56, 52, 52, 52, 51, 50, 50, 48, 48, 44, 43, 43, 41, 40, 40, 40, 36, 36, 35, 33, 31, 31, 30, 25, 20, 18, 13, 13, 10, 9, 8, 7, 7, 7, 6, 5, 3, -11, -11, -12, -13, -14, -15, -20, -23, -23, -23, -33, -33, -35, -35, -37, -37, -38, -43, -43, -44, -46, -47, -54, -55, -55, -59, -64, -64, -65, -67, -68, -70, -72, -74, -76, -79, -89, -90, -96, -96, -98]


**6.1** Then replace each value in Xsort with index i as the sum of the values with index i-1 and i. <br> 

In [21]:
# Regenerate Xsort by taking the first element and, for each subsequent index, sum the current element with the previous element
Xsort = [Xsort[0]] + [Xsort[i - 1] + Xsort[i] for i in range(1, len(Xsort))]

In [22]:
# Print the Xsort list
print("Xsort:", Xsort)

Xsort: [99, 197, 195, 193, 191, 186, 181, 177, 174, 173, 171, 170, 170, 166, 162, 158, 150, 140, 129, 123, 121, 117, 113, 108, 104, 104, 103, 101, 100, 98, 96, 92, 87, 86, 84, 81, 80, 80, 76, 72, 71, 68, 64, 62, 61, 55, 45, 38, 31, 26, 23, 19, 17, 15, 14, 14, 13, 11, 8, -8, -22, -23, -25, -27, -29, -35, -43, -46, -46, -56, -66, -68, -70, -72, -74, -75, -81, -86, -87, -90, -93, -101, -109, -110, -114, -123, -128, -129, -132, -135, -138, -142, -146, -150, -155, -168, -179, -186, -192, -194]


## Exercise II - The Most Uncool Band

**7**: Consider the dictionaries  *songs* and *users* that are declared in the cell below. <br>
Create a list with the names of the people who listened to more than one song. Print the List. <br>

In [23]:
songs = {
    "4589": "Mr. Blue Sky, Calling America, Twilight, Night in the City",
    "7312": "Strange Magic, Sweet Talkin' Woman, Turn to Stone",
    "9641": "Confusion, Last Train to London, All Over the World, Hold On Tight",
    "2853": "Can't Get It Out of My Head, Showdown, Twilight",
    "6704": "Ticket to the Moon, Here Is the News, Twilight, Secret Messages, Four Little Diamonds"
}

In [24]:
users = {
    "4589": "Eve",
    "7312": "Diana",
    "9641": "Alice",
    "2853": "Charlie",
    "6704": "Bob"
}

In [25]:
# Create a list to store the names of users who listened to more than one song
users_with_multiple_songs = []

# Iterate over the songs dictionary
for user_id, song_list in songs.items():
    # Split the "song_list" into individual songs and count the total number of songs
    songs_count = len(song_list.split(", "))
    
    # Check if the user listened to more than one song
    if songs_count > 1:
        # Store the user name associated with the user ID
        user_name = users[user_id]

        # Append the user name to the list "users_with_multiple_songs"
        users_with_multiple_songs.append(user_name)

In [26]:
# Print the answer  
print(f"The following users have listened to more than one song: {users_with_multiple_songs}.")

The following users have listened to more than one song: ['Eve', 'Diana', 'Alice', 'Charlie', 'Bob'].


**7.1** What is the name of the user who listened to the most songs?

*Tip: You will want to check the methods associated with string manipulation. See the link: <br>
https://python-reference.readthedocs.io/en/latest/docs/unicode/index.html*

In [27]:
# Initialize variables to store the maximum song count and a list for users with that count
max_songs = 0
users_max_songs = []

# Iterate over each user in the "songs" dictionary
for user_id, song_list in songs.items():
    # Split the "song_list" into individual songs and count the total number of songs
    songs_count = len(song_list.split(", "))
    
    # Check if the current user has listened to more songs than the current maximum
    if songs_count > max_songs:
        # Update the maximum song count with the current user's count
        max_songs = songs_count
        
        # Update the user with the maximum songs to the current user
        users_max_songs = [users[user_id]]

    # If the current user's song count equals the current maximum
    elif songs_count == max_songs:
        # Add the user to the list of users with the maximum songs
        users_max_songs.append(users[user_id])   

In [28]:
# Print the answer      
if len(users_max_songs) == 1:
    print(f"The user who listened to the most songs is {users_max_songs[0]}, with {max_songs} songs.")
else:
    print(f"The users who listened to the most songs ({max_songs} songs) are: {', '.join(users_max_songs)}.")

The user who listened to the most songs is Bob, with 5 songs.


**7.2** What is the name of the most popular song?

In [29]:
# Create a dictionary to store the count of each song
songs_count = {}

# Iterate over the lists of songs in the "songs" dictionary
for song_list in songs.values():
    # Split the song list into individual songs
    song_list_split = song_list.split(", ")
    
    # Iterate through each song in the "song_list_split"
    for song in song_list_split:
        # Check if the song is already in the "songs_count" dictionary
        if song in songs_count:
            # If the song exists, increment its count by 1
            songs_count[song] += 1
        else:
            # If the song does not exist, initialize its count to 1
            songs_count[song] = 1

# Find the maximum count of any song
max_songs_count = max(songs_count.values())

# Find all the songs that have the maximum count
most_popular_songs = [song for song, count in songs_count.items() if count == max_songs_count]

In [30]:
# Print the answer      
if len(most_popular_songs) == 1:
    print(f"The most popular song is '{most_popular_songs[0]}', with {max_songs_count} occurrences.")
else:
    print(f"The most popular songs ({max_songs_count} occurrences) are: {', '.join(most_popular_songs)}.")

The most popular song is 'Twilight', with 3 occurrences.


# <span style="color:brown"> Part 2 -  Data loading and analysis  </span>

## Exercise I - Alice what do you have to say?

**8**: Load the Alice text file into a variable called Alice. <br>

*Note: Use a relative filepath in relation to the location of your notebook.*

In [31]:
# Open the file "alice.txt" in read mode ("r")
with open("alice.txt", "r") as text_file:
    # Load the content of the text file into a variable
    Alice = text_file.read()

In [32]:
# Print the Alice variable
print(Alice)

Alice was beginning to get very tired of sitting by her sister on the bank, and of having nothing to do: once or twice she had peeped into the book her sister was reading, but it had no pictures or conversations in it, <<and what is the use of a book>>. thought Alice <<without pictures or conversation?>>.


**9**: Create a list in which each element is a word from the file *Alice*. Store that list in a variable called *wAlice*. <br>

*Note: You will need to do some text parsing here. In particular to split the sentences into words. It is also a good practice to normalize words so that words "Hello" and "hello" become identical, by making all letters lower case. <br>*

*Tip: check the following links for a discussion on regular expressions. Also consult the methods available for string manipulation.*

https://docs.python.org/3/library/re.html

https://stackoverflow.com/questions/1276764/stripping-everything-but-alphanumeric-chars-from-a-string-in-python

In [33]:
# Import the re module
import re

In [34]:
# Convert the "Alice" text to lowercase
Alice_lower = Alice.lower()

# Use a regular expression to replace all the non-alphanumeric characters with spaces
# re is the module for regular expressions, and sub stands for substitution
# [^a-z0-9\s] matches any character that is not a lowercase letter, digit, or space and replace it with spaces
Alice_sub = re.sub(r'[^a-z0-9\s]', ' ', Alice_lower)

# Split the "Alice_sub" text into words by using spaces as the separator
wAlice = Alice_sub.split()

In [35]:
# Print wAlice
print("wAlice:", wAlice)

wAlice: ['alice', 'was', 'beginning', 'to', 'get', 'very', 'tired', 'of', 'sitting', 'by', 'her', 'sister', 'on', 'the', 'bank', 'and', 'of', 'having', 'nothing', 'to', 'do', 'once', 'or', 'twice', 'she', 'had', 'peeped', 'into', 'the', 'book', 'her', 'sister', 'was', 'reading', 'but', 'it', 'had', 'no', 'pictures', 'or', 'conversations', 'in', 'it', 'and', 'what', 'is', 'the', 'use', 'of', 'a', 'book', 'thought', 'alice', 'without', 'pictures', 'or', 'conversation']


##### Using the list *wAlice* answer the following questions:<br>
**10**: How many words are in the file Alice.txt? <br>

In [36]:
# Print the answer
print(f"There are {len(wAlice)} words in the file 'Alice.txt'.")

There are 57 words in the file 'Alice.txt'.


**11**: What is the longest and smallest word in the text file?

*Note: Length in this case is measured in terms of the number of characters.*

In [37]:
# Find the longest word in the text file
longest_word = max(wAlice, key=len)

# Find the smallest word in the text file
smallest_word = min(wAlice, key=len)

In [38]:
# Print the answers
print(f"The longest word in the 'Alice.txt' file is '{longest_word}'.")
print(f"The smallest word in the 'Alice.txt' file is '{smallest_word}'.")

The longest word in the 'Alice.txt' file is 'conversations'.
The smallest word in the 'Alice.txt' file is 'a'.


**12**: Delete all the repeated words from *wAlice*. 

In [39]:
# Convert the list of words to a set to remove duplicates, as sets do not allow repeated elements
wAlice = set(wAlice)

# Convert the set back to a list to maintain list functionality
wAlice = list(wAlice)

In [40]:
# Print wAlice
print("wAlice:", wAlice)

wAlice: ['by', 'reading', 'conversation', 'pictures', 'she', 'what', 'had', 'no', 'once', 'having', 'was', 'tired', 'conversations', 'peeped', 'do', 'sister', 'in', 'on', 'the', 'it', 'thought', 'without', 'very', 'use', 'alice', 'into', 'twice', 'her', 'but', 'of', 'or', 'book', 'a', 'nothing', 'get', 'sitting', 'bank', 'and', 'beginning', 'is', 'to']


**12.1**:How many different words does the text contain?

In [41]:
# Print the answer
print(f"There are {len(wAlice)} unique words in the file 'Alice.txt'.")

There are 41 unique words in the file 'Alice.txt'.


## Exercise II - Of Countries I Love

Consider the list countries in the cell below. <br>
It consists of a list of the 3-digit ISO codes of a set of countries of interest.

**Note**: If you find the below list of tuples to be erroneous, correct it accordingly and comment on your correction.

In [42]:
countries = [
('PRT','Portugal','Europe'),
('ISL','Israel','Asia'),
('COL','Colombia','South America'),
('NEV','Nevada','North America'),
('JPN','Japan','Asia'),
('RUS','Russia','well... which part?'), 
('DEN','Denmark','Europe'),
('NOR','Norway','Europe')
]

**Now consider the file *cdata.csv* that you should download.** <br>

The file contains several information about countries, and is organized as follows:
1. Column 1 is the 3-digit **ISO Code**
2. Column 2 is the **Full Name** of the country
3. Column 3 is the **Continent** of the country
4. Column 4 is the **Population Size** in 2010
5. Column 5 is the **GDP per capita** in 2010

**13**: Using the Library CSV from Python STL, load the file *cdata.csv* into an object called *raw*. 

In [43]:
# Import the csv module
import csv

In [44]:
# Open the file "cdata.csv" in read mode ("r")
with open("cdata.csv", "r") as csv_file:
    # Use csv.reader to read the file and convert each row into a list of values
    # The result is a list of lists, where each inner list represents a row from the CSV file
    raw = list(csv.reader(csv_file))

In [45]:
# Print raw
print("raw:", raw)

raw: [['iso', 'countryname', 'continent', 'population(mil)', 'gdp'], ['ago', 'Angola', 'Africa', '23.369131088256836', '5988.534997149481'], ['bdi', 'Burundi', 'Africa', '8.766929626464844', '731.4232803265862'], ['ben', 'Benin', 'Africa', '9.199258804321289', '1919.9969479963038'], ['bfa', 'Burkina Faso', 'Africa', '15.605216979980469', '1327.2165314775982'], ['bwa', 'Botswana', 'Africa', '2.0148661136627197', '12256.14159052444'], ['caf', 'Central African Republic', 'Africa', '4.4485249519348145', '865.4421867263901'], ['cmr', 'Cameroon', 'Africa', '19.970495223999023', '2684.972890247272'], ['cod', 'Democratic Republic of the Congo', 'Africa', '64.52326202392578', '634.9713963013496'], ['cog', 'Republic of the Congo', 'Africa', '4.386693000793457', '4979.990496069269'], ['com', 'Comoros', 'Africa', '0.6896920204162598', '2292.997969425929'], ['cpv', 'Cape Verde', 'Africa', '0.5023840069770813', '5644.821253339075'], ['dji', 'Djibouti', 'Africa', '0.8511459827423096', '2124.442062597

**Note:** Now that we've opened the CSV file, we can correct the provided list of countries based on the actual data from the file.

In [46]:
# Create an empty list to store the corrected countries
corrected_countries = []

# Iterate over each tuple in the provided "countries" list
for iso_code, country_name, continent in countries:
    # Initialize flag to track whether the country is found in the CSV
    country_found = False

    # Search for the correct country data in the "raw" data, skipping the header row (column names)
    for row in raw[1: ]:
        # Match country names
        if row[1].lower() == country_name.lower():
            # Set flag to True if the country is found
            country_found = True
            
            # Extract the correct ISO code and continent from the "raw" data
            correct_iso = row[0]
            correct_continent = row[2]
            
            # Append the corrected tuple to "corrected_countries" with the correct ISO and continent
            corrected_countries.append((correct_iso, country_name, correct_continent))
            
            # Exit the inner loop once the country is found
            break
    
    # If the country was not found in the "raw" data, print a warning message
    if not country_found:
        print(f"Warning: {country_name} not found in the CSV. Skipping this entry.")

# Overwrite the original "countries" list with the corrected entries
countries = corrected_countries

# Print the corrected countries list
print("\nCorrected 'countries' list:")
countries


Corrected 'countries' list:


[('prt', 'Portugal', 'Europe'),
 ('isr', 'Israel', 'Asia'),
 ('jpn', 'Japan', 'Asia'),
 ('rus', 'Russia', 'Asia'),
 ('dnk', 'Denmark', 'Europe'),
 ('nor', 'Norway', 'Europe')]

**Notes:** After reviewing the "cdata.csv" file, here’s what we found and fixed:
- Colombia was not present in the CSV, so it was removed from the list;
- Nevada was skipped, as expected, since it’s a U.S. state, not a country;
- The ISO code for Israel was incorrect in the original list (ISL). We updated it to the correct code, ISR;
- Russia's continent was missing, so we added it as Asia based on the CSV data;
- Denmark's ISO code was listed as DEN, but the correct code is DNK, which we updated;
- Lastly, since the CSV file used lowercase ISO codes, we ensured all ISO codes in the list were automatically converted to lowercase during the loop to maintain consistency.

**13.1**: Create a dictionary called *cData* in which the key corresponds to the 3-digit ISO Code and the value is a tuple with the information contained in the 2nd to the 5th of column of the cdata file.

In [47]:
# Initialize an empty dictionary to store country data
cData = {}

# Iterate over the rows in the "raw" data, skipping the header row (column names)
for row in raw[1: ]:
    # Get the 3-digit ISO code from the first column
    iso_code = row[0]

    # Get the information from the 2nd to the 5th columns and store it as a tuple
    info = tuple(row[1:5])

    # Add the ISO code and its corresponding information to the "cData" dictionary
    cData[iso_code] = info

In [48]:
# Print cData
print("cData:", cData)

cData: {'ago': ('Angola', 'Africa', '23.369131088256836', '5988.534997149481'), 'bdi': ('Burundi', 'Africa', '8.766929626464844', '731.4232803265862'), 'ben': ('Benin', 'Africa', '9.199258804321289', '1919.9969479963038'), 'bfa': ('Burkina Faso', 'Africa', '15.605216979980469', '1327.2165314775982'), 'bwa': ('Botswana', 'Africa', '2.0148661136627197', '12256.14159052444'), 'caf': ('Central African Republic', 'Africa', '4.4485249519348145', '865.4421867263901'), 'cmr': ('Cameroon', 'Africa', '19.970495223999023', '2684.972890247272'), 'cod': ('Democratic Republic of the Congo', 'Africa', '64.52326202392578', '634.9713963013496'), 'cog': ('Republic of the Congo', 'Africa', '4.386693000793457', '4979.990496069269'), 'com': ('Comoros', 'Africa', '0.6896920204162598', '2292.997969425929'), 'cpv': ('Cape Verde', 'Africa', '0.5023840069770813', '5644.821253339075'), 'dji': ('Djibouti', 'Africa', '0.8511459827423096', '2124.4420625972525'), 'dza': ('Algeria', 'Africa', '36.117637634277344', '1

**14**: Using *cData*, identify what is the most common Continent among the nations in the list *countries*.

In [49]:
# Extract ISO codes from the "countries" list
iso_codes_from_countries = [country[0].lower() for country in countries]

# Filter cData to only include the countries from "countries" list
filtered_cData = {iso_code: info for iso_code, info in cData.items() if iso_code in iso_codes_from_countries}

# Extract the continents from the "filtered_cData"
continents = [info[1] for info in filtered_cData.values()]

# Count the occurrences of each continent
continents_counts = {continent: continents.count(continent) for continent in set(continents)}

# Calculate the maximum count of any continent
max_count = max(continents_counts.values())

# Find all continents that have the maximum count
most_common_continents = [continent for continent, count in continents_counts.items() if count == max_count]

In [50]:
# Print the answer
if len(most_common_continents) == 1:
    print(f"The most common continent is {most_common_continents[0]} with {max_count} occurrences.")
else:
    print(f"There is a tie between the most common continents (with {max_count} occurrences): {', '.join(most_common_continents)}.")

There is a tie between the most common continents (with 3 occurrences): Asia, Europe.


**15**: Using *cData*, identify what is the the most populated nation in the list *countries*.

In [51]:
# Initialize variables to store the most populated country (no ties are expected) and its population
most_populated_country = None
max_population = 0

# Iterate through filtered_cData (created in the exercise 14.) to find the country with the highest population
for iso_code, info in filtered_cData.items():
    # Extract the population and convert it to float
    population = float(info[2])
    
    # Check if the current country's population is greater than the "max_population"
    if population > max_population:
        # Update "max_population" to the current country's population
        max_population = population
    
        # Update "most_populated_country" to the current country's name
        most_populated_country = info[0]

In [52]:
# Print the answer
print(f"The most populated country is {most_populated_country} with a population of {round(max_population, 2)} thousands.")

The most populated country is Russia with a population of 143.15 thousands.


**16**: Compare the average GDP per capita of the nations in the list *countries* with the average GDP of the countries in cdata file. What can you conclude?

**Note:** Although the exercise states to compare the average GDP per capita of the countries in the list "countries" with the average GDP (and not average GDP per capita) of the countries in the "cdata.csv" file, we will instead compare the averages of GDP per capita for a more meaningful analysis. This is because the "cdata.csv" file provides GDP per capita data, allowing us to assess the standard of living of the countries in the list relative to the remaining countries in the CSV file.

In [53]:
# Initialize an empty list to store GDP per capita of countries in the "countries" list, using "filtered_cData" (created in exercise 14.)
gdp_countries_list = []

# Iterate through each country in the "filtered_cData"
for iso_code, info in filtered_cData.items():
    # Get GDP per capita and convert it to float
    gdp_per_capita = float(info[3])
    
    # Append the GDP per capita to the "gdp_countries_list"
    gdp_countries_list.append(gdp_per_capita)

# Calculate the average GDP per capita for the countries in the "countries" list
average_gdp_countries_list = sum(gdp_countries_list) / len(gdp_countries_list)

# Initialize an empty list to store GDP per capita of all countries except those in the "countries" list
gdp_all_countries = []

# Iterate over the entire cData to include all countries not in the "corrected_countries" list (i.e., not in "filtered_cData")
for iso_code, info in cData.items():
    # If the country is not in "filtered_cData"
    if iso_code not in filtered_cData:
        # Get GDP per capita and convert it to float
        gdp_per_capita = float(info[3])
        
        # Append the GDP per capita to the "gdp_all_countries"
        gdp_all_countries.append(gdp_per_capita)

# Calculate the average GDP per capita for all countries in cData except those in the "countries" list
average_gdp_all_countries = sum(gdp_all_countries) / len(gdp_all_countries)

In [54]:
# Print the results
print(f"Average GDP per capita for countries in the 'countries' list: {round(average_gdp_countries_list, 2)}.")
print(f"Average GDP per capita for all countries in the 'cdata.csv' file (excluding those in the 'countries' list): {round(average_gdp_all_countries, 2)}.")

Average GDP per capita for countries in the 'countries' list: 38242.63.
Average GDP per capita for all countries in the 'cdata.csv' file (excluding those in the 'countries' list): 17274.61.


In [55]:
# Calculate the ratio of average GDP per capita between the countries in the list and the remaining countries in the 'cdata.csv' file
average_gdp_ratio = average_gdp_countries_list / average_gdp_all_countries

# Print the conclusion
if average_gdp_ratio == 1:
    print(f"The ratio is 1.\nThe average GDP per capita of the countries in the 'countries' list is exactly equal to that of the remaining countries in the 'cdata.csv' file.\nThis suggests that the countries in the list have a similar economic performance and standard of living to the global average represented by the other countries.")
elif average_gdp_ratio > 1:
    print(f"The ratio is {round(average_gdp_ratio, 2)}.\nThe average GDP per capita of the countries in the 'countries' list is {round(average_gdp_ratio, 2)} times higher than that of the remaining countries in the 'cdata.csv' file.\nThis suggests that the countries in the list tend to be wealthier, with a higher standard of living, compared to the other countries.")
else:
    print(f"The ratio is {round(average_gdp_ratio, 2)}.\nThe average GDP per capita of the countries in the 'countries' list is {round(1 / average_gdp_ratio, 2)} times lower than that of the remaining countries in the 'cdata.csv' file.\nThis suggests that the countries in the list have a lower standard of living or economic output, potentially representing developing or emerging economies, compared to the other countries.")

The ratio is 2.21.
The average GDP per capita of the countries in the 'countries' list is 2.21 times higher than that of the remaining countries in the 'cdata.csv' file.
This suggests that the countries in the list tend to be wealthier, with a higher standard of living, compared to the other countries.


# <span style="color:brown"> Part 3 - Functions hurt nobody </span>

## Exercise I - I hate math

##### Consider the following equation:

\begin{equation}
    y = 6x^2 + 3x + 2 + cos(x)
\end{equation}

**17**: Write a function called *f* that takes one argument, x, and returns y according to the equation above. *Note: You should use the module math to solve this exercise.*

In [56]:
# Import the math module
import math

# Import the Union type hint from the typing module
from typing import Union

In [57]:
# Define the function f
# The 'Union' type hint allows to specify that a function parameter can accept more than one type
def f(x: Union[int, float]) -> float:
    return 6 * x**2 + 3*x + 2 + math.cos(x)

**17.1**: Call the function for x = 2 and print the answer.

In [58]:
# Call the f function for x = 2
answer = f(2)

# Print the answer
print(answer)

31.583853163452858


##### Consider the following sequence of numbers:

\begin{equation}
    x_0 = 0;  \\
    x_{n} = x_{n-1} - n; ~~~ \textit{ if} ~~~ x_{n-1}-n > 0 ~~~ \textit{ and is not already in the sequence}\\
    x_{n} = x_{n-1} + n; ~~~ \textit{ otherwise} ~~~ ~~~ ~~~ ~~~ ~~ ~~~ ~~~ ~~~ ~~~ ~~~ ~ ~~~ ~~~ ~~~ ~~~ ~~~ ~~~ ~ ~~~ ~~~ ~~~~ ~~~ ~~~ ~~~ ~ \\
\end{equation}

**17.2**: Write a function that returns the nth digit of the above defined sequence. <br>

*Note: the above sequence is also known as the Recamán's sequence, and it was invented by Bernardo Recamán Santos (Bogotá, Colombia)*

In [59]:
# Define the function RecamanSequence
def RecamanSequence(n: int) -> int:
    """
    Returns the nth digit in Recamán's sequence.

    Args:
    - n (int): The 1-based index of the digit in the sequence to return. Must be a positive integer.

    Returns:
    - int: The nth digit of the Recamán's sequence.
    """
        
    # Raise an error if n is not positive
    if n <= 0:
        raise ValueError("n must be a positive integer!")
    
    # Initialize the sequence with 0 as the first digit
    sequence = [0]

    # Create the sequence until the nth digit
    for i in range(1, n):
        # Subtract the current index from the last digit
        value = sequence[i-1] - i
        
        # If the value is positive and not already in the sequence, append it
        if value > 0 and value not in sequence:
            sequence.append(value)
        # Otherwise, add the index to the last digit and append it
        else:
            sequence.append(sequence[i-1] + i)
    
    # Return the nth digit in the sequence
    return sequence[n-1]

In [60]:
# Test the RecamanSequence function by printing the first 10 digits of Recamán's sequence
for n in range(1, 11):
    print(f"x{n} = {RecamanSequence(n)}")

x1 = 0
x2 = 1
x3 = 3
x4 = 6
x5 = 2
x6 = 7
x7 = 13
x8 = 20
x9 = 12
x10 = 21


**17.3** Write a function named isPrime that takes an integer as input and outputs True if the number is prime and False if the number if not prime. 

In [61]:
# Define the function isPrime
def isPrime(x: int) -> bool:
    """
    Check if a number is prime.

    Args:
    - x (int): The number to check for primality. Must be a positive integer.

    Returns:
    - bool: True if x is prime, False otherwise.
    """
    
    # Raise an error if x is not positive
    if x <= 0:
        raise ValueError("x must be a positive integer!")
    
    # Handle the case for 1 separately (1 is not a prime number)
    if x == 1:
        return False
    
    # Iterate from 2 to x-1 to check for divisors
    for i in range(2, x):
        # If x is divisible by any number in this range, it's not a prime number
        if x % i == 0:
            return False
    
    # If no divisors were found, x is prime
    return True

**17.4** Test your function against some examples to show that it works as expected.

In [62]:
# Test the isPrime function by printing the primality of the numbers from 1 to 10
for i in range(1, 11):
    print(f"{i}: {isPrime(i)}")

1: False
2: True
3: True
4: False
5: True
6: False
7: True
8: False
9: False
10: False


## Exercise II - Monty Hall Problem


**18**: The Monty Hall Problem is a probability puzzle loosely based on the American television game show "Let's Make a Deal" and named after its original host, Monty Hall. The puzzle can be stated as follows

> Suppose you're on a game show, and you're given the choice of three doors: Behind one door is a car; behind the others, goats. You pick a door, say No. 1, and the host, who knows what's behind the doors, opens another door, say No. 3, which has a goat. He then says to you, "Do you want to pick door No. 2?" Is it to your advantage to switch your choice? <br>
+info: https://en.wikipedia.org/wiki/Monty_Hall_problem

While intuitively your guess might be that the chances of winning the car is 1/3 independently of the choice to switch doors or not. However, the odds of winning if you switch doors are 2/3 and greater than if you decide not to switch. While there are theoretical solutions to this problem that allow us to estimate the correct odds of winning in each scenario -- switching or not doors -- an alternative way to proof such outcome is through computer simulations. That is the goal of this exercise, to write the necessary components to simulate the Monty Hall problem and validate the theoretical results through simulations.

In this exercise, start by implementing a function that simulates one instance of the Monty Hall problem. 
In that sense, write a function called MontyHall that accepts **one argument**:
- A boolean (True/False, or 1/0) that specifies if the player switches doors or not (after the host has opened his door, which contains a goat).
The function should return True if the player wins the car or False if not.

Naturally, for the function to reproduce the contest it needs to consider and perform some actions, such as:
- Set up the game, that is, create the necessary variables to store the information about three doors and what they have behind, as well as an indicator to track the choice of the player.
- Given the initial choice of the player, simulate the opening of one of the two remaining doors that contains a goat.
- Given the last two doors left, simulate if the player wants to switch or not his/her choice given their strategy (defined by the input boolean argument of the function).
- Output the result, if the contestant correctly guesses which door hides the car or if not.

The function MontyHall should simulate one instance of the contest given the choice of the player (to switch or not).

**The goal is to understand, statistically, which action leads to the highest probability of winning.** <br>
*Note: You should use the library random to solve this exercise.*

In [63]:
# Import the random module
import random

In [64]:
# Define the function MontyHall
def MontyHall(switch: bool) -> bool:
    """
    Simulate one instance of the Monty Hall problem.

    Args:
    - switch (bool): True if the player wants to switch doors after the host opens a door, which contains a goat, False otherwise.

    Returns:
    - bool: True if the player correctly guesses which door hides the car, False otherwise.
    """

    # Initialize the doors: one with a car, two with a goat
    doors = ["car", "goat", "goat"]

    # Shuffle the doors randomly
    random.shuffle(doors)

    # Simulate the player randomly selecting one of the three doors
    player_door = random.randint(0, 2)

    # Identify which doors were not chosen by the player
    remaining_doors = [i for i in range(3) if i != player_door]
    
    # Simulate the host opening a door, which contains a goat
    host_door = next(i for i in remaining_doors if doors[i] == "goat")
    
    # Simulate if the player switches doors
    if switch:
        # Remove the door opened by the host
        remaining_doors.remove(host_door)
        
        # Switch to the remaining door
        player_door = remaining_doors[0]

    # Return True if the player's final choice contains the car
    return doors[player_door] == "car"

In [65]:
# Test the MontyHall function with the player choosing to switch doors
MontyHall(True)

False

**18.1** Now that you have a function that simulates an instance of the Monty Hall problem, implement an experiment where you repeat many times (thousands of times) for each of the two possible scenarios: player switching the door, and player not switching the door. Keep track of the results and estimate the frequencies of wins for each scenario and discuss if the results are inline with the theoretical odds.

In [66]:
# Define the function ExperimentMontyHall
def ExperimentMontyHall(n: int):
    """
    Run an experiment to simulate the Monty Hall problem n times for each of the two possible scenarios:
    player switching the door, and player not switching the door.
    Calculate the win percentage for both scenarios and print the results.
    
    Args:
    - n (int): The number of simulations to run for each scenario (switching or not switching).

    Returns:
    - None: The function prints the win percentages for both scenarios, but does not return any value.
    """
    
    # Initialize a list to store the results of the simulations for both scenarios
    scenarios_simulations = []

    # Simulate the two possible scenarios: switching or not switching
    for switch in [True, False]:
        # Initialize a list to store the result of each simulation for the current scenario
        simulations = []

        # Simulate the Monty Hall problem n times for the current scenario, and append the results to "simulations"
        for _ in range(n):
            simulations.append(MontyHall(switch))
        
        # Append the list of results for the current scenario to "scenarios_simulations"
        scenarios_simulations.append(simulations)
    
    # Calculate the win percentage when the player switches doors
    switching_wins_percentage = sum(scenarios_simulations[0]) / n * 100
    
    # Calculate the win percentage when the player doesn't switch doors
    not_switching_wins_percentage = sum(scenarios_simulations[1]) / n * 100

    # Print the results
    print("Wins percentage")
    print(f"Switching scenario: {round(switching_wins_percentage, 2)}%")
    print(f"Not switching scenario: {round(not_switching_wins_percentage, 2)}%")

In [67]:
# Test the ExperimentMontyHall function with 10,000 simulations for each scenario
ExperimentMontyHall(10000)

Wins percentage
Switching scenario: 66.73%
Not switching scenario: 33.29%


For the switching and not switching scenarios, the win percentages in the simulation are 66.73% and 33.29%, respectively. These results are very close to the theoretical odds of 2/3 and 1/3, respectively.<br>

**Conclusion:** The results from 10,000 simulations align well with the theoretical odds of the Monty Hall problem. If the player chooses to switch doors, they significantly increase their chance of winning (around 2/3 of the time). On the other hand, if they stick with their original choice, their chances of winning are much lower (around 1/3 of the time).

## Exercise III

**19**: Consider the list *A*, declared below. <br>
Use the function map() to change the values of A by adding 1 to each value if it is even and subtracting 1 from it otherwise.

**Note**: Carefully consider the values of list A and adapt your code accordingly to deal with number that are represented as strings.

In [68]:
A = [460,3347,3044,490,699,"1258",1804,973,2223,3416,"2879",1058,2915,2422,351,"1543",1020,208,643,"795",3337,"2585",471,2623,1077]

In [69]:
# For each element of A (x), convert it to an integer:
# - If x is even, add 1 to it
# - If x is odd, subtract 1 from it
A = list(map(lambda x: int(x)+1 if int(x)%2 == 0 else int(x)-1, A))

In [70]:
# Print the modified list A
print(A)

[461, 3346, 3045, 491, 698, 1259, 1805, 972, 2222, 3417, 2878, 1059, 2914, 2423, 350, 1542, 1021, 209, 642, 794, 3336, 2584, 470, 2622, 1076]


**20**: Create a list B with the same size of A, where each element is True if the associated value in list A is greater than 700, else is False.

In [71]:
# Create the list B
B = list(map(lambda x: x>700, A))

In [72]:
# Print the list B
print(B)

[False, True, True, False, False, True, True, True, True, True, True, True, True, True, False, True, True, False, False, True, True, True, False, True, True]


**21**: Create a list L that contains the Logarithm of base 10 of each value in A. <br>

*Note: You should use the module math to solve this exercise.*

In [73]:
# Import the math module
import math

In [74]:
# Create the list L
L = list(map(lambda x: math.log10(x), A))

In [75]:
# Print the list L
print(L)

[2.663700925389648, 3.5245259366263757, 3.4835872969688944, 2.6910814921229687, 2.843855422623161, 3.1000257301078626, 3.256477206241677, 2.9876662649262746, 3.346744054604849, 3.5336449787987627, 3.4590907896005865, 3.024895960107485, 3.4644895474339714, 3.384353414137506, 2.5440680443502757, 3.188084373714938, 3.0090257420869104, 2.3201462861110542, 2.807535028068853, 2.8998205024270964, 3.523226041965701, 3.4122925093230463, 2.6720978579357175, 3.4186326873540653, 3.0318122713303706]


**22**: How many numbers in A are greater than 1000? <br>

*Note: You should use the function filter().*

In [76]:
# Create a filtered list with the elements of A that are greater than 1000
numbers_above_1000 = list(filter(lambda x: x>1000, A))

In [77]:
# Print the length of the filtered list (the number of elements of A that are greater than 1000)
print(f"The number of elements of A that are greater than 1000 is: {len(numbers_above_1000)}.")

The number of elements of A that are greater than 1000 is: 16.


**23** Use the function isPrime, that you declared in 17.3, to identify which numbers from A are prime. Note that depending on your implementation of isPrime this task can take more or less time. Use filter to achieve this task.

In [78]:
# Create a filtered list with the elements of A that are prime
prime_numbers = list(filter(lambda x: isPrime(x), A))

In [79]:
# Print the filtered list (the elements of A that are prime)
print(f"The prime numbers in A are: {prime_numbers}.")

The prime numbers in A are: [461, 491, 1259, 2423, 1021].


### Congratulations, it is done!